tuf/ngclient/_internal/metadata_bundle.py

```diff
@@ -0,0 +1,461 @@
+>>> with open(root_path, "rb") as f:
+>>>     bundle = MetadataBundle(f.read())
+>>>
+>>> # update root from remote until no more are available
+>>> with download("root", bundle.root.signed.version + 1) as f:
+>>>     bundle.update_root(f.read())
+>>> # ...
+>>> bundle.root_update_finished()
+>>>
+>>> # load local timestamp, then update from remote
+>>> try:
+>>>     with open(timestamp_path, "rb") as f:
+>>>         bundle.update_timestamp(f.read())
+>>> except (RepositoryError, OSError):
+>>>     pass # failure to load a local file is ok
+>>>
+>>> with download("timestamp") as f:
+>>>     bundle.update_timestamp(f.read())
+>>>
+>>> # load local snapshot, then update from remote if needed
+>>> try:
+>>>     with open(snapshot_path, "rb") as f:
+>>>         bundle.update_snapshot(f.read())
+>>> except (RepositoryError, OSError):
+>>>     # local snapshot is not valid, load from remote
+>>>     # (RepositoryErrors here stop the update)
+>>>     with download("snapshot", version) as f:
+>>>         bundle.update_snapshot(f.read())
+
+TODO:
+ * exceptions are not final: the idea is that client could just handle
+   a generic RepositoryError that covers every issue that server provided
+   metadata could inflict (other errors would be user errors), but this is not
+   yet the case
+ * usefulness of root_update_finished() can be debated: it could be done
+   in the beginning of load_timestamp()...
+ * some metadata interactions might work better in Metadata itself
+ * Progress through Specification update process should be documented
+   (not sure yet how: maybe a spec_logger that logs specification events?)
+"""
+
+import logging
+from collections import abc
+from datetime import datetime
+from typing import Dict, Iterator, Optional
+
+from securesystemslib import hash as sslib_hash
+from securesystemslib import keys as sslib_keys
+
+from tuf import exceptions
+from tuf.api.metadata import Metadata, Root, Targets
+from tuf.api.serialization import DeserializationError
+
+logger = logging.getLogger(__name__)
+
+# This is a placeholder until ...
+# TODO issue 1306: implement this in Metadata API
+def verify_with_threshold(
+    delegator: Metadata, role_name: str, unverified: Metadata
+) -> bool:
+    """Verify 'unverified' with keys and threshold defined in delegator"""
+    role = None
+    keys = {}
+    if isinstance(delegator.signed, Root):
+        keys = delegator.signed.keys
+        role = delegator.signed.roles.get(role_name)
+    elif isinstance(delegator.signed, Targets):
+        if delegator.signed.delegations:
+            keys = delegator.signed.delegations.keys
+            # role names are unique: first match is enough
+            roles = delegator.signed.delegations.roles
+            role = next((r for r in roles if r.name == role_name), None)
+    else:
+        raise ValueError("Call is valid only on delegator metadata")
+
+    if role is None:
+        raise ValueError(f"Delegated role {role_name} not found")
```

Comment:
I do allow that on purpose, and I think the exception is correct: caller is trying to verify a role that isn't delegated by delegator.

Reply:
But it's a valid allowed situation to have Targets without delegations and thus call  `verify_with_threshold` which will throw that exception.
Then, you should add a check inside `update_targets` which doesn't call `verify_with_threshold` on empty delegations
or catch the exception.

Reply:
btw, we should polish this code in the PR for #1306 (I plan to do one soon) -- it's here just as a placeholder

Reply:
can you comment in the code where we should handle the exception because I am not sure what this refers to?

If there are no delegations, then how does the client end up calling update_targets() for a delegate?

Reply:
You can call it here: https://github.com/jku/tuf/blob/f057d57e6aa2cc1bad1a153343fc6335c9c5c11c/tuf/ngclient/_internal/metadata_bundle.py#L445 line 445 in your current changes.
What's interesting is that you aren't testing if we support `targets` with no `delegations` field.

The reason is that you don't call `update_delegated_targets()` in your tests with a target file without `delegations`.
I thought you should been testing this in `test_metadata_bundle.py` when updating `role2.json` like this:
```python
        with open(os.path.join(repo_dir, "role2.json"), "rb") as f:
            bundle.update_delegated_targets(f.read(), "role2", "role1")
```
but after a quick check, I found that `role2.json` had `delegations`, but `delegations` itself consists only of empty `keys` dict and `roles` list.
I asked if that was the intention when marking `delegations` as optional [here](https://github.com/theupdateframework/specification/pull/165#issuecomment-855858990).

When I removed `delegations` from `role2.json` I got: 
```
tuf.exceptions.UnsignedMetadataError: New role2 is not signed by role1
```

Reply:
> When I removed `delegations` from `role2.json` I got:
> ```
> tuf.exceptions.UnsignedMetadataError: New role2 is not signed by role1
> ```

Did you resign the metadata after modifying? If you didn't, this seems like the correct thing to do -- signature in role2 is no longer valid

I'm not saying there isn't a bug here (and we definitely should have better ways of modifying and resigning the data) but I just don't see it. The delegations in the `delegated` metadata are not processed (apart from what happens in `Metadata.from_bytes()`) during `update_delegated_targets()`. 

The delegations in the `delegator` metadata are processed -- but in that case calling `update_delegated_targets()` on metadata that does not have delegations would be a bug and would result in a ValueError.

Reply:
Yea, I had forgotten to resign the metadata. I did the experiment again. 

> The delegations in the delegated metadata are not processed at all (apart from what happens in Metadata.from_bytes()) during update_delegated_targets().
The delegations in the delegator metadata are processed -- but in that case calling update_delegated_targets() on metadata that does not have delegations would be a bug and would result in a ValueError.

Yes, you are right and I verify that after debugging it.
Probably there is no bug here and it's me spending too much time on optional attributes. :D